### PR TITLE
Add a `summary` table reporter

### PR DESCRIPTION
**Pull request by mildm8nnered**
Adds a new summary reporter, that reports the frequency of rule violations. 

For example, when run on SwiftLint itself, like so `swiftlint --enable-all-rules --reporter summary`, the current output would be:

```
+-----------------------------------------+--------+-------------+--------+----------+--------+------------------+-----------------+
| rule identifier                         | opt-in | correctable | custom | warnings | errors | total violations | number of files |
+-----------------------------------------+--------+-------------+--------+----------+--------+------------------+-----------------+
| explicit_type_interface                 | no     | no          | no     |    3,462 |      0 |            3,462 |             480 |
| explicit_acl                            | no     | no          | no     |    3,310 |      0 |            3,310 |             483 |
| indentation_width                       | no     | no          | no     |    1,610 |      0 |            1,610 |             305 |
| prefer_nimble                           | no     | no          | no     |      771 |      0 |              771 |              58 |
| multiline_arguments_brackets            | no     | no          | no     |      753 |      0 |              753 |             128 |
| type_contents_order                     | no     | no          | no     |      651 |      0 |              651 |             268 |
| explicit_top_level_acl                  | no     | no          | no     |      644 |      0 |              644 |             401 |
| required_deinit                         | no     | no          | no     |      543 |      0 |              543 |             268 |
| no_magic_numbers                        | no     | no          | no     |      535 |      0 |              535 |              90 |
| implicit_return                         | no     | no          | no     |      461 |      0 |              461 |             202 |
| no_extension_access_modifier            | no     | no          | no     |        0 |    380 |              380 |             207 |
| multiline_parameters_brackets           | no     | no          | no     |      250 |      0 |              250 |              68 |
| multiline_arguments                     | no     | no          | no     |      223 |      0 |              223 |              58 |
| anonymous_argument_in_multiline_closure | no     | no          | no     |      187 |      0 |              187 |              72 |
| file_types_order                        | no     | no          | no     |      169 |      0 |              169 |              82 |
| no_grouping_extension                   | no     | no          | no     |      165 |      0 |              165 |             152 |
| multiline_literal_brackets              | no     | no          | no     |      122 |      0 |              122 |              31 |
| vertical_whitespace_between_cases       | no     | no          | no     |      113 |      0 |              113 |              43 |
| force_unwrapping                        | no     | no          | no     |      101 |      0 |              101 |              35 |
| conditional_returns_on_newline          | no     | no          | no     |       83 |      0 |               83 |              54 |
| explicit_enum_raw_value                 | no     | no          | no     |       78 |      0 |               78 |              14 |
| trailing_closure                        | no     | no          | no     |       78 |      0 |               78 |              46 |
| missing_docs                            | no     | no          | no     |       69 |      0 |               69 |              13 |
| switch_case_on_newline                  | no     | no          | no     |       57 |      0 |               57 |              12 |
| convenience_type                        | no     | no          | no     |       52 |      0 |               52 |              48 |
| prefixed_toplevel_constant              | no     | no          | no     |       52 |      0 |               52 |              26 |
| multiline_parameters                    | no     | no          | no     |       51 |      0 |               51 |              32 |
| strict_fileprivate                      | no     | no          | no     |       51 |      0 |               51 |              21 |
| multiline_function_chains               | no     | no          | no     |       44 |      0 |               44 |              28 |
| discouraged_optional_collection         | no     | no          | no     |       25 |      0 |               25 |              15 |
| function_default_parameter_at_end       | no     | no          | no     |       17 |      0 |               17 |              12 |
| closure_body_length                     | no     | no          | no     |       14 |      1 |               15 |               6 |
| implicitly_unwrapped_optional           | no     | no          | no     |        6 |      0 |                6 |               5 |
| self_binding                            | no     | no          | no     |        4 |      0 |                4 |               3 |
| line_length                             | no     | no          | no     |        2 |      0 |                2 |               1 |
| empty_count                             | no     | no          | no     |        0 |      1 |                1 |               1 |
| function_body_length                    | no     | no          | no     |        1 |      0 |                1 |               1 |
+-----------------------------------------+--------+-------------+--------+----------+--------+------------------+-----------------+
| Total                                   |        |             |        |   14,754 |    382 |           15,136 |             538 |
+-----------------------------------------+--------+-------------+--------+----------+--------+------------------+-----------------+
```

`swiftlint --reporter summary` would produce

```
+-----------------+--------+-------------+--------+----------+--------+------------------+-----------------+
| rule identifier | opt-in | correctable | custom | warnings | errors | total violations | number of files |
+-----------------+--------+-------------+--------+----------+--------+------------------+-----------------+
| Total           |        |             |        |        0 |      0 |                0 |               0 |
+-----------------+--------+-------------+--------+----------+--------+------------------+-----------------+
```

The main use case for this is when you're configuring SwiftLint - rules with thousands or hundreds violations will pretty much have to be disabled, whereas rules with smaller numbers of violations are good candidates for fixing, or suppression.

### CODE REVIEW

File: Package.swift [hunk 1]
```diff
@@ -61,7 +61,7 @@ let package = Package(
         ),
         .target(
             name: "SwiftLintFramework",
-            dependencies: frameworkDependencies
+            dependencies: frameworkDependencies + ["SwiftyTextTable"]
```
Comment (mildm8nnered) [inline Package.swift:64]: Sadly, had to add this linker dependency :-(

File: Tests/SwiftLintFrameworkTests/ReporterTests.swift [hunk 2]
```diff
@@ -6,19 +6,20 @@ import XCTest
 class ReporterTests: XCTestCase {
     func testReporterFromString() {
         let reporters: [Reporter.Type] = [
-            XcodeReporter.self,
```
Comment (mildm8nnered) [inline Tests/SwiftLintFrameworkTests/ReporterTests.swift:9]: I re-ordered these alphabetically while I was here.

File: Package.swift (hunk 1 above)
Comment (mildm8nnered) [reply Package.swift:64]: Which I guess was not enough, as CI and `bazel build :swiftlint` are failing with 

```
error: emit-module command failed with exit code 1 (use -v to see invocation)
Source/SwiftLintFramework/Reporters/SummaryReporter.swift:2:8: error: no such module 'SwiftyTextTable'
import SwiftyTextTable
       ^
Source/SwiftLintFramework/Reporters/SummaryReporter.swift:2:8: error: no such module 'SwiftyTextTable'
import SwiftyTextTable
```

**Comment by mildm8nnered** [comment]
I just noticed https://github.com/realm/SwiftLint/pull/3661, which is an old and outdated pass at pretty much the same thing.

File: Package.swift (hunk 1 above)
Comment (mildm8nnered) [reply Package.swift:64]: OK - I think I've figured the bazel linking out

File: Source/SwiftLintFramework/Reporters/SummaryReporter.swift [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+
+        var totalNumberOfViolations = 0
+
+        for ruleIdentifier in sortedRuleIdentifiers {
+            guard let ruleIdentifier = ruleIdentifiersToViolationsMap[ruleIdentifier]?.first?.ruleIdentifier else {
+                continue
+            }
+
+            var rule: Rule?
+            if let ruleType = primaryRuleList.list[ruleIdentifier] {
```
Comment (SimplyDanny) [inline Source/SwiftLintFramework/Reporters/SummaryReporter.swift:56]: You can just `let rule = primaryRuleList.list[ruleIdentifier]` and then check `rule is OptInRule.Type`.

File: Source/SwiftLintFramework/Reporters/SummaryReporter.swift [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+                rule = ruleType.init()
+            }
+
+            let numberOfViolations = ruleIdentifiersToViolationsMap[ruleIdentifier]?.count ?? 0
+            totalNumberOfViolations += numberOfViolations
+            let ruleViolations = ruleIdentifiersToViolationsMap[ruleIdentifier] ?? []
+            let numberOfFiles = Set(ruleViolations.map { $0.location.file }).count
```
Comment (SimplyDanny) [inline Source/SwiftLintFramework/Reporters/SummaryReporter.swift:63]: Is the number of file such a useful information? 😅

**Review by SimplyDanny — COMMENTED** [review]
Looks good to me!

A pity there was no reply in #3661 anymore. Now, time was invested twice.

One important part in the other PR was the separation of errors and warnings. Do you think you can easily incorporate this aspect into you implementation?

File: Source/SwiftLintFramework/Reporters/SummaryReporter.swift (hunk 4 above)
Comment (mildm8nnered) [reply Source/SwiftLintFramework/Reporters/SummaryReporter.swift:63]: I think 1,000 violations in 1 file out of 1,000 is different from 1,000 violations in 1,000 files ...

**Comment by mildm8nnered** [comment]
I'll update this and try to get the number of warnings and errors in.

**Review by SimplyDanny — APPROVED** [review]

**Comment by agustr** [comment]
Wooo hooo
Im so happy I could contribute a tiny bit.

**Comment by agustr** [comment]
little sad my code never made it in though.

**Comment by SimplyDanny** [comment]
Next one is yours @agustr! 😄

**Pull request merged**